src/vivarium_census_prl_synth_pop/components/observers.py

```diff
@@ -494,3 +472,199 @@ def get_observation(self, event: Event) -> pd.DataFrame:
+
+        # increment income based on the job the simulant has during
+        # the course of the time_step, which might change if we do
+        # this check on_time_step instead of on_time_step__prepare
+        builder.event.register_listener("time_step__prepare", self.on_time_step__prepare)
+        self.income_to_date = empty_income_series()
+        self.time_step = builder.configuration.time.step_size  # in days
+
+    def on_time_step__prepare(self, event):
+        pop = self.population_view.get(
+            event.index,
+        )
+        pop["income"] = self.pipelines["income"](pop.index)
+
+        # increment income for all person/employment pairs with income > 0
+        income_this_time_step = pd.Series(
+            pop["income"].values * self.time_step / DAYS_PER_YEAR,
+            index=pd.MultiIndex.from_arrays(
+                [pop.index, pop["employer_id"]], names=["simulant_id", "employer_id"]
+            ),
+        )
+
+        income_this_time_step = income_this_time_step[income_this_time_step > 0]
+
+        self.income_to_date = self.income_to_date.add(income_this_time_step, fill_value=0.0)
+
+    def to_observe(self, event: Event) -> bool:
+        """Only observe if this is the final time step of the sim"""
+        tax_date = dt.datetime(event.time.year, 1, 1)
+        return self.clock() < tax_date <= event.time
+
+    def get_observation(self, event: Event) -> pd.DataFrame:
+        pop = self.population_view.get(
+            event.index,
+        )
+        pop["income"] = self.pipelines["income"](pop.index)
+
+        ### create dataframe of all person/employment pairs
+
+        # start with income to date, which has simulant_id and employer_id as multi-index
+        self.income_to_date.name = "income"  # HACK: it would be nice if this name stayed
+        # with the pd.Series, but it is getting lost at some point in the computation
+
+        df_w2 = self.income_to_date.reset_index()
+        df_w2["tax_year"] = event.time.year
+
+        # merge in simulant columns based on simulant id
+        for col in [
+            "first_name_id",
+            "middle_name_id",
+            "last_name_id",
+            "age",
+            "date_of_birth",
+            "sex",
+            "ssn",
+            "address_id",
+            "housing_type",
+        ]:
+            df_w2[col] = df_w2["simulant_id"].map(pop[col])
+
+        # for simulants without ssn, record a simulant_id for a random household
+        # member with an ssn, if one exists
+
+        simulants_wo_ssn = pd.Series(
+            df_w2[~df_w2["ssn"]].index, index=df_w2[~df_w2["ssn"]].index
+        )
+        household_members_w_ssn = (
+            pop[pop["ssn"]].groupby("address_id").apply(lambda df_g: list(df_g.index))
+        )
+        household_members_w_ssn = simulants_wo_ssn.map(household_members_w_ssn).dropna()
+
+        ssn_for_simulants_wo = household_members_w_ssn.map(self.np_randomness.choice)
```

Comment:
Why not use vivarium's randomness stream?

Reply:
I couldn't figure out how to do this with a vivarium randomness method. patches welcome!